### PR TITLE
add CtKey and NatKey to avoid ambiguities

### LINKED ISSUE #8306: cilium: panic: interface conversion: tuple.TupleKey is *ctmap.CtKey4Global, not *tuple.TupleKey4Global
Potentially related to 79c247a955f8c3e02f0 (/cc @aanm):
```
[...]
level=info msg="Envoy: Starting xDS gRPC server listening on /var/run/cilium/xds.sock" subsys=envoy-manager
level=info msg="Validating configured node address ranges" subsys=daemon
level=info msg="Starting connection tracking garbage collector" subsys=daemon
level=warning msg="Garbage collection on ipv4 TCP CT map failed to finish" interrupted=0 subsys=map-ct
panic: interface conversion: tuple.TupleKey is *ctmap.CtKey4Global, not *tuple.TupleKey4Global

goroutine 252 [running]:
github.com/cilium/cilium/pkg/maps/nat.(*Map).DeleteMapping(0xc00010b800, 0x3312e00, 0xc000a982e0, 0x0, 0x0)
	/root/go/src/github.com/cilium/cilium/pkg/maps/nat/nat.go:236 +0x133
github.com/cilium/cilium/pkg/maps/ctmap.purgeCtEntry4(0xc000a94120, 0x3312e00, 0xc000a982e0, 0xc00010b800, 0xc00075f8ac, 0x4)
	/root/go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:317 +0xc2
github.com/cilium/cilium/pkg/maps/ctmap.doGC4.func1(0x32fd3a0, 0xc000a982e0, 0x32e15a0, 0xc000987140)
	/root/go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:350 +0x247
github.com/cilium/cilium/pkg/bpf.(*Map).DumpReliablyWithCallback(0xc000a94120, 0xc00075fae0, 0xc0001082d0, 0x0, 0x0)
	/root/go/src/github.com/cilium/cilium/pkg/bpf/map_linux.go:697 +0x833
github.com/cilium/cilium/pkg/maps/ctmap.doGC4(0xc000a94120, 0xc0000c55e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:382 +0x228
github.com/cilium/cilium/pkg/maps/ctmap.doGC(0xc000a94120, 0xc0000c55e0, 0x0)
	/root/go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:415 +0xd5
github.com/cilium/cilium/pkg/maps/ctmap.GC(0xc000a94120, 0xc0000c55e0, 0xc000a94120)
	/root/go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:430 +0x3a
github.com/cilium/cilium/pkg/endpointmanager.runGC(0x0, 0xc0000d0001, 0xc0000c55e0, 0x9, 0x0)
	/root/go/src/github.com/cilium/cilium/pkg/endpointmanager/conntrack.go:68 +0x49d
github.com/cilium/cilium/pkg/endpointmanager.EnableConntrackGC.func1(0xc000a98294, 0xc000480001, 0xc0000d0140, 0x1, 0x1, 0xc000a98298, 0xc000a1ac60)
	/root/go/src/github.com/cilium/cilium/pkg/endpointmanager/conntrack.go:118 +0x196
created by github.com/cilium/cilium/pkg/endpointmanager.EnableConntrackGC
	/root/go/src/github.com/cilium/cilium/pkg/endpointmanager/conntrack.go:113 +0xf4
```

### PR DESCRIPTION
**Pull request by aanm**
Fixes #8306

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8314)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

File: pkg/maps/ctmap/types.go [hunk 1]
```diff
@@ -104,6 +108,356 @@ func (m MapType) isTCP() bool {
...
+	DestAddr   types.IPv4      `align:"daddr"`
+	SourceAddr types.IPv4      `align:"saddr"`
+	DestPort   uint16          `align:"dport"`
+	SourcePort uint16          `align:"sport"`
+	NextHeader u8proto.U8proto `align:"nexthdr"`
+	Flags      uint8           `align:"flags"`
+}
+
+// NewValue creates a new bpf.MapValue.
+func (k *CtKey4) NewValue() bpf.MapValue { return &CtEntry{} }
```
Comment (borkmann) [inline pkg/maps/ctmap/types.go:136]: Could we make `NewValue` be part of the interface to make it clear that CT and NAT map have to override this, or technically not possible? Iirc, with the TupleKey back then, is was that both derived their types from tuple key and hence both had separate `NewValue` helpers. Would we otherwise need to cast incompatible value types when accessing them after lookup (CT vs NAT value if default is the former)?

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: cilium/cmd/bpf_ct_flush.go [hunk 2]
```diff
@@ -44,6 +46,26 @@ type dummyEndpoint struct {
 	ID int
 }
 
+func init() {
+	natGlobalMaps := nat.GlobalMaps(true, true)
+	global4Map := natGlobalMaps[0]
+	global6Map := natGlobalMaps[1]
+
+	// SNAT also only works if the CT map is global so all local maps will be nil
+	ctmap.InitMapInfo(option.Config.CTMapEntriesGlobalTCP, option.Config.CTMapEntriesGlobalAny,
```
Comment (joestringer) [inline cilium/cmd/bpf_ct_flush.go]: Why not pass `natGlobalMaps` into `InitMapInfo`, avoid the explicit map declaration in multiple places when the details of the map declared in the argument here are all ctmap-specific? (This is now very specific detail cloned in two places).

File: cilium/cmd/bpf_nat_list.go [hunk 3]
```diff
@@ -51,7 +51,7 @@ func dumpNat() {
 		}
 		if err != nil {
 			if err == os.ErrNotExist {
-				Fatalf("Unable to open %s: %s: please try using \"cilium bpf nat list\"", path, err)
+				fmt.Printf("Unable to open %s: %s: please try using \"cilium bpf nat list\"", path, err)
```
Comment (joestringer) [inline cilium/cmd/bpf_nat_list.go]: Errors should go to stderr.

File: cilium/cmd/bpf_nat_flush.go [hunk 4]
```diff
@@ -46,7 +46,7 @@ func flushNat() {
 			err = m.Open()
 		}
 		if err != nil {
-			Fatalf("Unable to open %s: %s", path, err)
+			fmt.Printf("Unable to open %s: %s", path, err)
```
Comment (joestringer) [inline cilium/cmd/bpf_nat_flush.go]: Errors should go to stderr. Also, is this related to the other changes?

File: daemon/daemon.go [hunk 5]
```diff
@@ -666,8 +667,28 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
...
+	natGlobalMaps := nat.GlobalMaps(option.Config.EnableIPv4, option.Config.EnableIPv6)
+	var global4Map, global6Map ctmap.NatMap
+	if option.Config.EnableIPv4 {
+		global4Map = natGlobalMaps[0]
+	}
+	if option.Config.EnableIPv6 {
+		global6Map = natGlobalMaps[1]
+	}
+
+	// SNAT also only works if the CT map is global so all local maps will be nil
```
Comment (joestringer) [inline daemon/daemon.go]: Why does SNAT only work if the ct map is global?

Also, this is a functional change which would really be nice to review as a separate commit to make it easier to ensure that all the right places are touched to cover a new nil case.

File: pkg/maps/ctmap/lookup.go [hunk 6]
```diff
@@ -20,83 +20,9 @@ import (
 	"strconv"
 
 	"github.com/cilium/cilium/pkg/bpf"
-	"github.com/cilium/cilium/pkg/byteorder"
-	"github.com/cilium/cilium/pkg/tuple"
 	"github.com/cilium/cilium/pkg/u8proto"
 )
 
-// CtKey4 is needed to provide CtEntry type to Lookup values
```
Comment (joestringer) [inline pkg/maps/ctmap/lookup.go:28]: Please don't delete huge swathes of code and place it into another file in the same commit as functional changes.

File: pkg/maps/nat/nat.go [hunk 7]
```diff
@@ -183,43 +183,47 @@ func (m *Map) Flush() int {
...
+		CtKey6Global: *ctKey,
+	}
 	// Workaround #5848.
 	addr := key.SourceAddr
 	key.SourceAddr = key.DestAddr
 	key.DestAddr = addr
 	valMap, err := m.Lookup(&key)
 	if err == nil {
-		val := *(*NatEntry6)(unsafe.Pointer(valMap.GetValuePtr()))
+		val := (*NatEntry6)(valMap.GetValuePtr())
```
Comment (joestringer) [inline pkg/maps/nat/nat.go]: Is this no longer dereferencing the pointer because there is one less level of nesting in the structure?

**Review by joestringer — DISMISSED** [review]
A few minor nits around stderr for CLI, and that map complexity should be kept inside the map package.

Do we need to backport the panic fix? Big refactors should be kept separate if backports are necessary.

It would also be really nice to keep code moves in their own commits so it's easier to see what's supposed to change or not. The first commit starts in this direction, but then the second commit has a bunch of fairly significant code moves within the same package, alongside functional changes.

**Review by aanm — COMMENTED** [review]
@joestringer I kind of addressed your comments by re-writing everything. Needs a new review :-)

File: daemon/daemon.go (hunk 5 above)
Comment (aanm) [reply daemon/daemon.go]: @borkmann Can give more details here

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/maps/ctmap/ctmap.go [hunk 8]
```diff
@@ -92,6 +92,12 @@ const (
 	metricsDeleted = "deleted"
 )
 
+type NatMap interface {
```
Comment (brb) [inline pkg/maps/ctmap/ctmap.go:95]: Can we move this declaration to `pkg/maps/nat/types.go`?

**Review by brb — COMMENTED** [review]
LGTM, just one nit.

File: pkg/maps/ctmap/ctmap.go (hunk 8 above)
Comment (aanm) [reply pkg/maps/ctmap/ctmap.go:95]: Won't defeat the purpose of having an interface? :-)

Comment (joestringer) [reply pkg/maps/ctmap/ctmap.go:95]: Yeah, interfaces should be declared in the package where they're received, like it is here. I think it's fine.

**Comment by borkmann** [comment]
test-me-please

File: daemon/daemon.go (hunk 5 above)
Comment (borkmann) [reply daemon/daemon.go]: @joestringer My impression was that local CT got deprecated long ago. We could implement it if we think it makes sense to do so. Do we have users running local per container CT? In that case, daemon would need to GC a stand-alone CT map for bpf_netdev.

Comment (joestringer) [reply daemon/daemon.go]: I believe it's still there, though it doesn't get a lot of attention. I'm not aware of any users using it.

The main reason I ask is that this PR makes changes like:

```
	setupMapInfo(MapType(MapTypeIPv4TCPLocal), "CT_MAP_TCP4",		setupMapInfo(MapType(MapTypeIPv4TCPLocal), "CT_MAP_TCP4",
		&CtKey4{}, int(unsafe.Sizeof(CtKey4{})),			&CtKey4{}, int(unsafe.Sizeof(CtKey4{})),
		MapNumEntriesLocal, natV4)
```
=> 
```
	setupMapInfo(MapType(MapTypeIPv4TCPLocal), "CT_MAP_TCP4",
		&CtKey4{}, int(unsafe.Sizeof(CtKey4{})),			&CtKey4{}, int(unsafe.Sizeof(CtKey4{})),
		MapNumEntriesLocal, natV4)			MapNumEntriesLocal, natMaps[MapTypeIPv4TCPLocal])
```

Where in the previous code, the `natMaps[]` reference referred to the global map, while in the new version it is nil. I haven't validated what the subsequent behaviour is if these pointers are set to nil, and given that the change isn't called out in the commit/PR, I wasn't sure if this was intentional or not.

Comment (borkmann) [reply daemon/daemon.go]: Makes sense, in that case if BPF based SNAT is enabled, the code should bail out and not allow local maps to be configured by the daemon.

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
only failure was https://github.com/cilium/cilium/issues/8346

**Review by borkmann — APPROVED** [review]
lgtm afaict

**Comment by borkmann** [comment]
test-me-please
